goal_src/kernel/gstate.gc

```diff
@@ -85,56 +85,84 @@ There are several ways to "go"
   `(inspect-process-tree *active-pool* 0 0 ,detail)
   )
 
+;; use a compile-time list to keep track of the type of an anonymous behavior.
+(seval (define *defstate-type-stack* '()))
+(desfun def-state-check-behavior (beh-form beh-type)
+  "check if code block is an anonymous behavior. needed for anonymous behaviors on defstate."
 
-;; define a state state
-(defmacro defstate (state-name
+  (when (and (pair? beh-form) (eq? (first beh-form) 'behavior))
+    (push! *defstate-type-stack* beh-type)
+    )
+  )
+
+(defmacro defstate (state-name parents
                     &key (event #f)
                     &key (enter #f)
                     &key (trans #f)
                     &key (exit #f)
                     &key (code #f)
                     &key (post #f)
                       )
-  `(begin
-    (define ,state-name (new 'static 'state
-                             :name (quote ,state-name)
-                             :next #f
-                             :exit #f
-                             :code #f
-                             :trans #f
-                             :post #f
-                             :enter #f
-                             :event #f
-                             )
+  "Define a new state!"
+
+  (with-gensyms (new-state)
+    (let ((defstate-type (first parents)))
+      (def-state-check-behavior event defstate-type)
```

Comment:
Should we `(set! *defstate-type-stack* '())` at the beginning?  If you hit a syntax error in the middle of compiling a behavior it will stop compiling immediately and leave stuff in the stack.

Reply:
I totally forgot about that. I had it set up differently originally and forgot about this when I changed things, I definitely want it to be cleared every time.